Fearure/increment3

Orel-AI opened the pull request.

Orel-AI committed:
first increment

Orel-AI committed:
added test and refactor

Orel-AI committed:
added main test file

Orel-AI committed:
added main test file

Orel-AI committed:
added main test file

Orel-AI committed:
reduced violations

Orel-AI committed:
reduced violations (Close Request's Body in handler_test.go)

Orel-AI committed:
added chi http library

Orel-AI committed:
reduced violations after static analyzer

lugingf commented on storage/storage.go:
```diff
@@ -0,0 +1,22 @@
+}
+
+func AddRecord(key string, data string, ctx context.Context) {
+	linkMap[key] = data
```
Тут не критично, но на будущее расскажу штуку:
При большой нагрузке конкурентная запись в мапу выстрелит паникой. Чтобы избежать можно или мютексом блочить (не всегда помогает), либо еще есть штука `sync.Map`

lugingf commented on cmd/shortener/main.go:
```diff
@@ -1,4 +1,19 @@
+	r.Post("/", handler.GenerateShorterLinkPOST)
+	err := http.ListenAndServe(":8080", r)
+	if err != nil {
+		panic(err)
```
давай не `panic()`, давай `log.Fatal()` :)

lugingf commented on service/shortener/service.go:
```diff
@@ -0,0 +1,53 @@
+	value := storage.FindRecord(encodedString, ctx)
+	if value == link {
+		return encodedString, nil
+	} else if value != "" && value != link {
```
`else` не нужны, давай от них избавимся, структура будет попроще

lugingf commented on service/shortener/service.go:
```diff
@@ -0,0 +1,53 @@
+	"net/url"
+)
+
+func GetShortLink(link string, ctx context.Context) (string, error) {
```
Вот тут тоже хочется структуру сервиса, и чтобы эти функции были методами этого сервиса. А сервис обращался к стораджу

lugingf commented on api/handler/handler.go:
```diff
@@ -0,0 +1,61 @@
+	"strings"
+)
+
+func GenerateShorterLinkPOST(w http.ResponseWriter, r *http.Request) {
```
Давай сделаем структуру хэндлера, в которую положим нужные нам ресурсы (сервис, а в сервис - сторадж). И эти функции станут методами структуры хэндлера

lugingf reviewed (changes requested):
В общем все хорошо. Оставил комментариев чуть-чуть. Переделывать не много, просто чуть привести к порядку структуру

Orel-AI committed:
refactor after review
-added structs for storage, service, handler
-changed storage from map[] to sync.Map
-in main.go changed logging of error
-reduced some logic in service.go

lugingf reviewed (approved):
Отличненько :)
